File: php/src/ConstantTime.php
<?php

declare(strict_types=1);

namespace SecureKit;

/**
 * Timing-attack-resistant comparison utility.
 */
final class ConstantTime
{
    public static function equals(string $a, string $b): bool
    {
        // hash_equals is constant-time for equal-length inputs and safely
        // short-circuits (in constant relative time) on length mismatch,
        // same trade-off as the Go/other ports: lengths of tokens aren't secret.
        return hash_equals($a, $b);
    }
}

File: go/constanttime.go
package securekit

import "crypto/subtle"

// ConstantTimeEqual compares two byte slices in constant time, regardless of
// where the first difference occurs, to resist timing side-channel attacks.
// It returns true only if the slices have equal length and identical content.
func ConstantTimeEqual(a, b []byte) bool {
	if len(a) != len(b) {
		// Length mismatch is intentionally allowed to short-circuit: comparing
		// two different-length secrets never happens in real CSRF/token flows
		// (tokens have a fixed encoded length), and hiding this doesn't add
		// meaningful protection while len() itself is not secret.
		return false
	}
	return subtle.ConstantTimeCompare(a, b) == 1
}

// ConstantTimeEqualString is a string convenience wrapper around ConstantTimeEqual.
func ConstantTimeEqualString(a, b string) bool {
	return ConstantTimeEqual([]byte(a), []byte(b))
}
